src/V3Premit.cpp

```diff
@@ -285,81 +271,56 @@ class PremitVisitor final : public VNVisitor {
         checkNode(nodep);
     }
     void visit(AstSel* nodep) override {
-        iterateAndNextNull(nodep->fromp());
+        iterate(nodep->fromp());
```

Comment:
Many places you've downgraded to iterate. In general I prefer always using iterateAndNextNull even when in a specific node type it won't have a next, because it sets ourselves up otherwise for a copy-and-paste error when moved into a new visitor that does have a next.  The single extra not taken branch on null nextp() seems a worthy price.

We probably could teach astgen to make custom "iterateAndNextNull" overloads for all AstNode flavors that have no next and just call iterate() instead.  Then no cut-and-paste concern and might be faster globally (or might be loss due to splitting hotness between two functions?)

Reply:
> The single extra not taken branch on null nextp() seems a worthy price.

There is lot more than a single extra not taken branch there. `iterate` is just `nodep->accept(visitor)`. `iterateAndNextNull` calls `AstNote::iterateteAndNext`, which does the whole dance setting up niterp/looping/etc.

Reply:
> We probably could teach astgen to make custom "iterateAndNextNull" overloads for all AstNode flavors that have no next

Whether a type of node has a next depends on the other nodes that hold it as child (via a List[_]), not the node itself, so I'm not sure what you are suggesting here. Maybe you are thinking about generating `iterateChildren` as a virtual function, which I have tried, and is a net even to a very small positive performance effect.

Reply:
Ah, right.  Doesn't address my concern about copy-and-paste trap though, and adds to the idea for what I described as a general improvement.

Reply:
Comment above was to two comments ago ;)

For the example shown nodep->fromp() is an AstNodeExpr.  If we teach astgen that AstNodeExpr can have no nextp (assuming that's true, need to check), then this can statically declare "iterateAndNext(AstNodeExpr*) as iterate()".

Reply:
BTW another possibility is to have iterateAndNext know what node types have nextp(), and self-simplify.  Perhaps we assign node type identifiers with even/odd bit 0 meaning if they have nextp().

Reply:
BTW another possibility is to have iterateAndNext know what node types have nextp(), and self-simplify at runtime based on nodep->nodeTypeCanHaveANext.  Perhaps to create nodeTypeCanHaveANext and have it fast we assign node type identifiers with even/odd bit 0 meaning if they have nextp().

Reply:
> BTW another possibility is to have iterateAndNext know what node types have nextp(), and self-simplify at runtime based on nodep->nodeTypeCanHaveANext. Perhaps to create nodeTypeCanHaveANext and have it fast we assign node type identifiers with even/odd bit 0 meaning if they have nextp().

Or `VNVisitor::iterateAndNext` can just be `if (nodep->nextp()) nodep->iterateAndNext(this) else node->accept(this)`